carla/recourse_methods/catalog/actionable_recourse/model.py

```diff
@@ -0,0 +1,212 @@
+        coeffs : np.ndArray
+            Coefficients
+        intercepts
+        """
+        self._data = mlmodel.data
+        self._mlmodel = mlmodel
+
+        # normalize and encode data
+        self._norm_enc_data = scale(
+            mlmodel.scaler, self._data.continous, self._data.raw
+        )
+        self._norm_enc_data = encode(
+            mlmodel.encoder, self._data.categoricals, self._norm_enc_data
+        )
+
+        # Get hyperparameter
+        self._fs_size = (
+            100 if "fs_size" not in hyperparams.keys() else hyperparams["fs_size"]
+        )
+        self._discretize_continuous = (
+            False
+            if "discretize" not in hyperparams.keys()
+            else hyperparams["discretize"]
+        )
+        self._sample_around_instance = (
+            True if "sample" not in hyperparams.keys() else hyperparams["sample"]
+        )
+
+        # Build ActionSet
+        self._action_set = rs.ActionSet(
+            X=self._norm_enc_data[self._mlmodel.feature_input_order]
+        )
+
+        # transform immutable feature names into encoded feature names of self._data.encoded_normalized
+        self._immutables = encoded_immutables(
+            self._mlmodel.data.immutables, self._mlmodel.feature_input_order
+        )
+
+        for feature in self._immutables:
+            self._action_set[feature].mutable = False
+            self._action_set[feature].actionable = False
+
+        self._coeffs, self._intercepts = coeffs, intercepts
+
+    def get_lime_coefficients(self, factuals):
+        """
+        Actionable Recourse is only defined on linear models. To make it work for arbitrary non-linear networks
+        we need to find the lime coefficients for every instance.
+
+        Parameters
+        ----------
+        factuals : pd.DataFrame
+            Instances we want to get lime coefficients
+
+        Returns
+        -------
+        coeffs : np.ndArray
+        intercepts : np.ndArray
+
+        """
+        coeffs = np.zeros(factuals.shape)
+        intercepts = []
+        lime_data = self._norm_enc_data[self._mlmodel.feature_input_order]
+        lime_label = self._norm_enc_data[self._data.target]
+
+        lime_exp = LimeTabularExplainer(
+            training_data=lime_data.values,
+            training_labels=lime_label,
+            feature_names=self._mlmodel.feature_input_order,
+            discretize_continuous=self._discretize_continuous,
+            sample_around_instance=self._sample_around_instance,
+            categorical_names=[
+                cat
+                for cat in self._mlmodel.feature_input_order
+                if cat not in self._data.continous
+            ]
+            # self._data.encoded_normalized's categorical features contain feature name and value, separated by '_'
+            # while self._data.categoricals do not contain those additional values.
+        )
+
+        for i in range(factuals.shape[0]):
```

Comment:
Can we also use iterrows here? Also as a sidenote, looping over dataframes is considered very bad practice, we could also think of using .apply() here or something else more elegant: https://stackoverflow.com/a/55557758